### PR TITLE
Delete Command: Add multi index delete and delete person with specific tags

### LINKED ISSUE #6: As a student with a small circle of friends, I can delete groups of contacts
...so that I will not have too many info of people I no longer contact

### LINKED ISSUE #81: 🐞 Prefix `-n` for name still required in `add` command
### Describe the issue
*What is expected to happen? How can it be reproduced?*

Entering `add John` command raises invalid command error, currently still `add -n John`.
Should be the former according to the [user guide](https://ay2021s2-cs2103-w16-3.github.io/tp/UserGuide.html).

________________________________________________________________
### Additional context
*Suggest possible bug sources (or [reference them](https://docs.github.com/en/github/managing-your-work-on-github/opening-an-issue-from-code)). Add screenshots if applicable.*

image

________________________________________________________________
### Environment
 - OS: Windows 10
 - JDK version: 11

### PR DESCRIPTION
**Pull request by garyljj**
Closes #6 
Fixes #81

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -67,15 +67,20 @@ Format: `clear`
...
+  * All indexes refers to the index number shown in the displayed person list (without sorting).
+  * All indexes must be a positive integer valid in the list.
+* If provided with tags
+  * Delete every person with the given tag.
+  * If the person is tagged by another tag, only person's tag is removed.
 
 Examples:
 * `delete 3` deletes contact at 3rd index.
 * `delete 3 4 5` deletes contacts at 3rd, 4th and 5th index.
+* `delete -t colleague` deletes contact with tag "colleague".
```
Comment (Ellevy) [inline docs/UserGuide.md:82]:
```suggestion
* `delete -t colleague` deletes the "colleague" tag in contacts that contains this tag. If "colleague" is the only tag that that contact contains, the contact will be deleted.
```

File: docs/UserGuide.md [hunk 2]
```diff
@@ -67,15 +67,20 @@ Format: `clear`
...
-* The index refers to the index number shown in the displayed person list (without sorting).
-* The index must be a positive integer valid in the list.
+Format: `delete {INDEX [INDEX]... | -t TAG [-t TAG]...}`
+* If provided with index(es)
+  * Deletes the person at the specified `INDEX`.
+  * All indexes refers to the index number shown in the displayed person list (without sorting).
+  * All indexes must be a positive integer valid in the list.
+* If provided with tags
+  * Delete every person with the given tag.
+  * If the person is tagged by another tag, only person's tag is removed.
```
Comment (Ellevy) [inline docs/UserGuide.md:77]:
```suggestion
  * Delete every person that has only 1 tag which is the given tag.
  * If the person is tagged by another tag, only the given tag is removed from the person. The contact will not be deleted.
```

File: docs/UserGuide.md [hunk 3]
```diff
@@ -67,15 +67,20 @@ Format: `clear`
...
 
 Examples:
 * `delete 3` deletes contact at 3rd index.
 * `delete 3 4 5` deletes contacts at 3rd, 4th and 5th index.
+* `delete -t colleague` deletes contact with tag "colleague".
+* `delete -t colleague -t cs2103` deletes contacts with tag "colleague" and contacts with tag "cs2103"
```
Comment (Ellevy) [inline docs/UserGuide.md:83]:
```suggestion
* `delete -t colleague -t cs2103` deletes the "colleague" tag and "cs2103" tag in contacts that contains these tags. If "colleague" and "cs2103" are the only tags that that contact contains, the contact will be deleted.
```

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 4]
```diff
@@ -1,53 +1,17 @@
...
+public abstract class DeleteCommand extends Command {
 
     public static final String COMMAND_WORD = "delete";
 
     public static final String MESSAGE_USAGE = COMMAND_WORD
-            + ": Deletes the person identified by the index number used in the displayed person list.\n"
-            + "Parameters: INDEX (must be a positive integer)\n"
-            + "Example: " + COMMAND_WORD + " 1";
+            + ": Deletes the person identified by the index number used in the displayed person list\n"
+            + "OR delete the person with the provided tag.\n"
```
Comment (Ellevy) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:12]:
```suggestion
            + "OR delete the person if that person has only 1 tag which is the provided tag\n"
            + "OR remove the provided tag from the person if that person has other tags.\n"
```

File: docs/UserGuide.md (hunk 3 above)
Comment (pyuxiang) [reply docs/UserGuide.md:83]: This is a little confusing. By deleting the contact when no more tags exists, it suggests that contacts cannot have no tags. But what about a workflow where user wants to delete all tags, then assign new tags from scratch?

Should the `delete -t` command just deal with tag deletion alone, and not touch contact deletion at all?

Comment (Ellevy) [reply docs/UserGuide.md:83]: Does this mean we should create another command just to handle contact deletion and this `delete -t` command only deals with tag deletion? Basically similar to our `add`, `find` and `delete` commands that only deal with contacts while `tags` command only deals with tags.

Comment (pyuxiang) [reply docs/UserGuide.md:83]: I think using the current `add`, `find`, `delete` for contact processing is a good idea, like what you mentioned. That's essentially the existing commit: "deletes all contacts with at least one of colleague or cs2103 tags".

It's the proposed change that I don't quite catch, i.e. "delete tags in contact + delete contact if there is an exact tag match".

PS: On this note, going by your suggestion to use `tags` exclusively for tag, we can do tag deletion using the same command, e.g. `tags -d -t colleague -t cs2103`. Sounds like a pretty self-consistent idea.

Comment (garyljj) [reply docs/UserGuide.md:83]: Maybe is just the phrasing confusing. So in the latest commit I phrased the main info as
* If provided with tags
  * Delete every person who is tagged with the specified tag.
  * If the person is tagged with another tag, only the specified tag will be removed. The contact will not be deleted.
 
 Then for these examples, I left it as
" `delete -t colleague -t cs2103` deletes contacts with tag "colleague" and contacts with tag "cs2103" "
Cos the happy path is to delete the contact. Then ONLY IF it is tagged with another tag then the contact is not deleted.

File: docs/UserGuide.md (hunk 1 above)
Comment (garyljj) [reply docs/UserGuide.md:82]: Did not change, discussion is below

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java (hunk 4 above)
Comment (garyljj) [reply src/main/java/seedu/address/logic/commands/DeleteCommand.java]: Edited but phrased a little differently, check latest commit

**Review by pyuxiang — APPROVED** [review]
LGTM

**Review by Ellevy — APPROVED** [review]
LGTM

**Pull request merged**